### ISSUE TITLE
Pytorch Build: Caffe2 libcaffe2_gpu.so undefined reference to '__cudaPopCallConfiguration'

### ISSUE THREAD

**Issue by fwillo** [pytorch/pytorch#15433]
Dear all,

I try to build PyTorch from sources because I want to use CUDA aware OpenMPI with Infiniband support in Distributions. I pulled PyTorch from the Git repository und checked it out with `git checkout tags/v1.0.0` as well as initalized the submodules it with `git submodule update --init`. OpenMPI builds correctly with no errors. 
PyTorch build is called with
```
TORCH_CUDA_ARCH_LIST="6.1;6.0;5.2;3.5" USE_CUDNN=1 USE_DISTRIBUTED=1 USE_FFMPEG=1 NO_MKLDNN=1 
CUDA_HOME=$CUDA_ROOT CFLAGS='-O3 -g' python setup.py install
```
In my environment I set `LD_LIBRARY_PATH`, `CPATH` and `LIBRARY_PATH` correctly, so that they point to the following software packages
```
slurm/17.02.4
gcc/4.8.5
cuda/9.1.85
nccl/2.2.13-cuda-9.2
cudnn/9.2-v7.4.1.5
gdrcopy/1.3-cuda9.1
mellanox_hpc-x/4.3-1.0.1.0/hcoll-cuda
mellanox_hpc-x/4.3-1.0.1.0/mxm
mellanox_hpc-x/4.3-1.0.1.0/ucx-cuda
mellanox_hpc-x/4.3-1.0.1.0/knem
zlib/1.2.8
ffmpeg/3.0
openmpi/gcc/3.0.0-test
```
Unfortunately I don't have the lines of error (exactly), but it happens after the successfull build of libcaffe2_gpu.so, where test programs are build, that the build aborts with the error message, that libcaffe2_gpu.so has an undefined reference to `__cudaPopCallConfiguration` and `__cudaPushCallConfiguration`. Looking at the linkage of libcaffe2_gpu.so with ldd shows:
```
$ ldd -r /tmp/pytorch/1.0.0/build/lib/libcaffe2_gpu.so | grep -i -E 'cuda|nccl'
	libc10_cuda.so => /tmp/pytorch/1.0.0/build/lib/libc10_cuda.so (0x00007f3e6cc9d000)
	libcusparse.so.9.1 => /usr/local/cuda/9.1.85/lib64/libcusparse.so.9.1 (0x00007f3e62b7d000)
	libcurand.so.9.1 => /usr/local/cuda/9.1.85/lib64/libcurand.so.9.1 (0x00007f3e5ebfa000)
	libcudnn.so.7 => /usr/local/cudnn/cuda9.2-v7.4.1.5/lib64/libcudnn.so.7 (0x00007f3e4a598000)
	libcudart.so.9.1 => /usr/local/cuda/9.1.85/lib64/libcudart.so.9.1 (0x00007f3e49cbe000)
	libcufft.so.9.1 => /usr/local/cuda/9.1.85/lib64/libcufft.so.9.1 (0x00007f3e425b2000)
	libcublas.so.9.1 => /usr/local/cuda/9.1.85/lib64/libcublas.so.9.1 (0x00007f3e3ee91000)
	libhcoll.so.1 => /usr/local/mellanox_hpc-x/4.3-1.0.1.0/hcoll/cuda/lib/libhcoll.so.1 (0x00007f3e3e0e8000)
	libocoms.so.0 => /usr/local/mellanox_hpc-x/4.3-1.0.1.0/hcoll/cuda/lib/libocoms.so.0 (0x00007f3e3dc87000)
	libalog.so.1 => /usr/local/mellanox_hpc-x/4.3-1.0.1.0/hcoll/cuda/lib/libalog.so.1 (0x00007f3e3d4fc000)
	libnccl.so.2 => /usr/local/nccl/2.2.13-cuda-9.2/lib/libnccl.so.2 (0x00007f3e38bfa000)
	libcuda.so.1 => /usr/lib64/libcuda.so.1 (0x00007f3e35eec000)
	libgdrapi.so.1 => /usr/local/gdrcopy/1.3-cuda9.1/lib64/libgdrapi.so.1 (0x00007f3e35ce8000)
	libicudata.so.58 => /usr/local/miniconda/envs/miniconda-ompi-test/lib/././libicudata.so.58 (0x00007f3e30b9a000)
undefined symbol: __cudaPushCallConfiguration	(/tmp/pytorch/1.0.0/build/lib/libcaffe2_gpu.so)
undefined symbol: __cudaPopCallConfiguration	(/tmp/pytorch/1.0.0/build/lib/libcaffe2_gpu.so)
```
Some research on the internet shows, that NCCL might be the problem. As already noticed, in my current build environment libnccl is successfully acknowledged and loaded. Looking for the symbols in libnccl.so, specifically for the CallConfiguration functions, I get:
```
$ readelf -a /usr/local/nccl/2.2.13-cuda-9.2/lib/libnccl.so | grep -i callconfig
  2579: 000000000006de40   285 FUNC    LOCAL  DEFAULT   10 __cudaPushCallConfigurati
  2695: 0000000000069a30   207 FUNC    LOCAL  DEFAULT   10 __cudaPopCallConfiguratio
```
So I can ensure, that libnccl.so is loaded in libcaffe2_gpu.so and libnccl.so contains the functions on which there are undefined references to. I don't understand where the problem is. Everything should be there, or isn't it? I'm pretty sure to miss out a crucial point here.

Another thing I noticed is the following.
```
$ ldd -r /usr/local/nccl/2.2.13-cuda-9.2/lib/libnccl.so.2
undefined symbol: dlsym	(/usr/local/nccl/2.2.13-cuda-9.2/lib/libnccl.so.2)
undefined symbol: dlerror	(/usr/local/nccl/2.2.13-cuda-9.2/lib/libnccl.so.2)
undefined symbol: dlopen	(/usr/local/nccl/2.2.13-cuda-9.2/lib/libnccl.so.2)
undefined symbol: dlclose	(/usr/local/nccl/2.2.13-cuda-9.2/lib/libnccl.so.2)
	linux-vdso.so.1 =>  (0x00007ffe10425000)
	librt.so.1 => /usr/lib64/librt.so.1 (0x00007f9f24c94000)
	libstdc++.so.6 => /usr/local/gcc/4.8.5/lib64/libstdc++.so.6 (0x00007f9f24992000)
	libm.so.6 => /usr/lib64/libm.so.6 (0x00007f9f24690000)
	libgcc_s.so.1 => /usr/local/gcc/4.8.5/lib64/libgcc_s.so.1 (0x00007f9f2447a000)
	libc.so.6 => /usr/lib64/libc.so.6 (0x00007f9f240ad000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9f2979e000)
	libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007f9f23e91000)
```
The NCCL library is also missing symbols from libdl (although I downloaded the pre-compiled libraries from NVidia). Maybe this causes the problem?

Looking forward to your answers, I'm running out of ideas. :(

Best wishes,
fwillo

**Comment by ckyleda** [comment]
I also have this error:
/libcaffe2_gpu.so: undefined reference to `__cudaPushCallConfiguration'
/libcaffe2_gpu.so: undefined reference to `__cudaPopCallConfiguration'

**Comment by fwillo** [comment]
@ckyleda Here is the archive name I downloaded from the official Nvidia web site: nccl_2.2.13-1+cuda9.2_x86_64.txz

Can you confirm this?

**Comment by ckyleda** [comment]
@fwillo 

Yes, it seems that the cuda 9.2 nccl libraries include those calls.

However, I'm running with Cuda 9.0 and the respective nccl libraries packaged with PyTorch, so why am I seeing this error?

I suspect the PyTorch devs have screwed something up.

**Comment by fwillo** [comment]
I just noticed that I used the CUDA 9.2 version of NCCL and now tried it with 2.1.15-cuda9.1. Same result as before. Maybe the __cudaPopCallConfiguration is somewhere else?

**Comment by ckyleda** [comment]
@fwillo 

Are those calls present in that version  of nccl? Must be something added after Cuda 9, and then they incorrectly assumed that all versions of NCCL include this.

Do you still see the linking errors inside the libnccl.so file?

**Comment by fwillo** [comment]
For 2.1.15-cuda9.1 I also have```

undefined symbol: dlsym	(/usr/local/nccl/2.1.15-cuda-9.1/lib/libnccl.so.2)
undefined symbol: dlerror	(/usr/local/nccl/2.1.15-cuda-9.1/lib/libnccl.so.2)
undefined symbol: dlopen	(/usr/local/nccl/2.1.15-cuda-9.1/lib/libnccl.so.2)
undefined symbol: dlclose	(/usr/local/nccl/2.1.15-cuda-9.1/lib/libnccl.so.2)
```

**Comment by ckyleda** [comment]
I noticed that I had installed a potentially incorrect version of 'magma-cuda'.

I'd installed magma-cuda92 rather than magma-cuda90. Currently building to see if this helps, though I don't hold out much hope. 

Might be worth checking.

**Comment by fwillo** [comment]
@ckyleda I found the cudatoolkit==7.0 installed in my environment and the order of pathes in LD_LIBRARY_PATH, LIBRARY_PATH, CPATH and PATH was the way, that my environment was favoured to the cuda installation in /usr/local/cuda. This wasn't the problem.

I also saw magma-cuda91 and magma-cuda92 installed. Because of your hint I deinstalled magma-cuda92 (to keep it 9.1 pure). It also appears to be working for me. I feel pretty stupid now.

Thanks for the hint @ckyleda and sorry for bothering everyone! Have nice holidays.
